### PR TITLE
new feature store and ground truth notebook

### PR DESCRIPTION
**Pull request by paul-cb**
*Issue #, if available:*

*Description of changes:*
This notebook demonstrates how to securely store the output of an image or text classification labelling job from Amazon Ground Truth directly into Feature Store using KMS key.

*Testing done:*

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

- [x] I have read the [CONTRIBUTING](https://github.com/aws/amazon-sagemaker-examples/blob/master/CONTRIBUTING.md) doc and adhered to the example notebook best practices
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/aws/amazon-sagemaker-examples/blob/master/README.md)
- [x] I have tested my notebook(s) and ensured it runs end-to-end
- [ ] I have linted my notebook(s) and code using `tox -e black-format,black-nb-format`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: sagemaker-featurestore/feature_store_object_detection_ground_truth.ipynb [hunk 1]
```diff
@@ -0,0 +1,497 @@
+{
```
Comment (can-sun) [inline sagemaker-featurestore/feature_store_object_detection_ground_truth.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #2.    s3_bucket_name = sagemaker_session.default_bucket() #This is the bucket for your Feature Store.
</pre><p>I would prefer to replace Feature Store with Offline Store</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/2822/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (can-sun) [inline sagemaker-featurestore/feature_store_object_detection_ground_truth.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #20.    #download_file_from_s3(public_s3_bucket_name, path='PATH', filename='output.manifest')
</pre><p>Should we uncomment this line?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/2822/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

Comment (can-sun) [inline sagemaker-featurestore/feature_store_object_detection_ground_truth.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #14.        for entry in open(filename, 'r'):
</pre><p> </p><p>Is it possible to rename and clarify what i, k, j means here? I've never used GroundTruth before, so these symbols look confusing to me. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/2822/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (can-sun) [inline sagemaker-featurestore/feature_store_object_detection_ground_truth.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #43.    df = create_dataframe_from_manifest('data/output.manifest')
</pre><p>Should we provide a default path here?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/2822/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (paul-cb) [reply sagemaker-featurestore/feature_store_object_detection_ground_truth.ipynb:1]: Fixed.

Comment (paul-cb) [reply sagemaker-featurestore/feature_store_object_detection_ground_truth.ipynb:1]: Actually, the`output.manifest` file is stored in the data folder so it's the default path.

**Review by can-sun — DISMISSED** [review]

**Review by alex-tang — DISMISSED** [review]

File: sagemaker-featurestore/feature_store_object_detection_ground_truth.ipynb [hunk 2]
```diff
@@ -0,0 +1,484 @@
...
+    "\n",
+    "This notebook uses KMS key for server side encryption for your Feature Store. For more information on server-side encryption, see [Feature Store: Encrypt Data in your Online or Offline Feature Store using KMS key](https://sagemaker-examples.readthedocs.io/en/latest/sagemaker-featurestore/feature_store_kms_key_encryption.html). \n",
+    "\n",
+    "If you would like to encrypt your data on the client side prior to ingestion, see [Amazon SageMaker Feature Store: Client-side Encryption using AWS Encryption SDK](https://sagemaker-examples.readthedocs.io/en/latest/sagemaker-featurestore/feature_store_client_side_encryption.html) for a demonstration. \n",
```
Comment (jsspric) [inline sagemaker-featurestore/feature_store_object_detection_ground_truth.ipynb:20]: "To encrypt your data on the client side prior to ingestion, see [Amazon SageMaker Feature Store: Client-side Encryption using AWS Encryption SDK](https://sagemaker-examples.readthedocs.io/en/latest/sagemaker-featurestore/feature_store_client_side_encryption.html) for a demonstration. \n",

File: sagemaker-featurestore/feature_store_object_detection_ground_truth.ipynb [hunk 3]
```diff
@@ -0,0 +1,484 @@
...
+    "\n",
+    "### Overview\n",
+    "1. Set up\n",
```
Comment (jsspric) [inline sagemaker-featurestore/feature_store_object_detection_ground_truth.ipynb:23]: Add period to be consistent with other list items.

File: sagemaker-featurestore/feature_store_object_detection_ground_truth.ipynb [hunk 4]
```diff
@@ -0,0 +1,484 @@
...
+    "2. Prepare `output.manifest`for Feature Store. \n",
```
Comment (jsspric) [inline sagemaker-featurestore/feature_store_object_detection_ground_truth.ipynb:24]: Missing space after "manifest"

File: sagemaker-featurestore/feature_store_object_detection_ground_truth.ipynb [hunk 5]
```diff
@@ -0,0 +1,484 @@
...
+    "3. Create a feature group and ingest your data into it.\n",
+    "\n",
+    "### Prerequisites\n",
+    "This notebook uses the Python SDK library for Feature Store, and the `Python 3 (Data Science)` kernel. To encrypt your data with KMS key for server side encryption, you will need to have an active KMS key. If you do not have a KMS key, then you can create one by following the [KMS Policy Template](https://sagemaker-examples.readthedocs.io/en/latest/sagemaker-featurestore/feature_store_kms_key_encryption.html#KMS-Policy-Template) steps or you can visit the [KMS section in the console](https://console.aws.amazon.com/kms/home) and follow the prompts for creating a KMS key. This notebook is compatible with SageMaker Studio, Jupyter, and JupyterLab. \n",
```
Comment (jsspric) [inline sagemaker-featurestore/feature_store_object_detection_ground_truth.ipynb]: "This notebook uses the Python SDK library for Feature Store, and the `Python 3 (Data Science)` kernel. To encrypt your data with a KMS key for server-side encryption, you must have an active KMS key. If you do not have a KMS key, then you can create one by following the [KMS Policy Template](https://sagemaker-examples.readthedocs.io/en/latest/sagemaker-featurestore/feature_store_kms_key_encryption.html#KMS-Policy-Template) steps or you can visit the [KMS section in the console](https://console.aws.amazon.com/kms/home) and follow the prompts for creating a KMS key. This notebook is compatible with SageMaker Studio, Jupyter, and JupyterLab. \n",

File: sagemaker-featurestore/feature_store_object_detection_ground_truth.ipynb [hunk 6]
```diff
@@ -0,0 +1,484 @@
...
+    "\n",
+    "### Library Dependencies:\n",
```
Comment (jsspric) [inline sagemaker-featurestore/feature_store_object_detection_ground_truth.ipynb]: "### Library dependencies:\n"

File: sagemaker-featurestore/feature_store_object_detection_ground_truth.ipynb [hunk 7]
```diff
@@ -0,0 +1,484 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "sagemaker_session = sagemaker.Session()\n",
+    "s3_bucket_name = sagemaker_session.default_bucket() # This is the bucket for your Feature Store.\n",
```
Comment (jsspric) [inline sagemaker-featurestore/feature_store_object_detection_ground_truth.ipynb]: "s3_bucket_name = sagemaker_session.default_bucket() # This is the bucket for your feature store.\n",

File: sagemaker-featurestore/feature_store_object_detection_ground_truth.ipynb [hunk 8]
```diff
@@ -0,0 +1,484 @@
...
+    "prefix = 'sagemaker-featurestore-demo'\n",
+    "role = sagemaker.get_execution_role()\n",
+    "region = sagemaker_session.boto_region_name"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Additional - Helper Method\n",
```
Comment (jsspric) [inline sagemaker-featurestore/feature_store_object_detection_ground_truth.ipynb]: "### Helper method\n",

File: sagemaker-featurestore/feature_store_object_detection_ground_truth.ipynb [hunk 9]
```diff
@@ -0,0 +1,484 @@
...
+    "Below is a method that you can use to get your manifest file from your S3 bucket into your current working directory."
```
Comment (jsspric) [inline sagemaker-featurestore/feature_store_object_detection_ground_truth.ipynb:80]: "Use the following method to get your manifest file from your S3 bucket into your current working directory."

File: sagemaker-featurestore/feature_store_object_detection_ground_truth.ipynb [hunk 10]
```diff
@@ -0,0 +1,484 @@
...
+    "    '''\n",
+    "    import os.path\n",
+    "    if not os.path.exists(filename):\n",
+    "        s3 = boto3.client('s3')\n",
+    "        s3.download_file(\n",
+    "            Bucket = bucket,\n",
+    "            Key = path,\n",
+    "            Filename = filename\n",
+    "        )\n",
+    "# Supply the path to your output.manifest file from your AWS Ground Truth Classification labelling job. \n",
```
Comment (jsspric) [inline sagemaker-featurestore/feature_store_object_detection_ground_truth.ipynb]: "# Supply the path to your `output.manifest` file from your Amazon SageMaker Ground Truth classification labeling job. \n",

File: sagemaker-featurestore/feature_store_object_detection_ground_truth.ipynb [hunk 11]
```diff
@@ -0,0 +1,484 @@
...
+    "#download_file_from_s3(public_s3_bucket_name, path='PATH/TO/GROUND-TRUTH/CLASSIFICATION-JOB/', filename='output.manifest')"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Prepare your manifest file for Feature Store. \n",
+    "\n",
+    "Below is a method that will parse your `output.manifest` file into a Pandas data frame for ingestion into your Feature Store. At this point, it is assumed that your manifest file is in your current working directory. "
```
Comment (jsspric) [inline sagemaker-featurestore/feature_store_object_detection_ground_truth.ipynb]: "Use the following method to parse your `output.manifest` file into a Pandas data frame for ingestion into your feature store. At this point, we assume that your manifest file is in your current working directory. "

File: sagemaker-featurestore/feature_store_object_detection_ground_truth.ipynb [hunk 12]
```diff
@@ -0,0 +1,484 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def create_dataframe_from_manifest(filename):\n",
+    "    '''\n",
+    "    filename: path to your output.manifest file. This should be the output.manifest file from an AWS Ground Truth Classification labelling job. \n",
```
Comment (jsspric) [inline sagemaker-featurestore/feature_store_object_detection_ground_truth.ipynb]: "filename: path to your `output.manifest` file. This should be the output.manifest file from an Amazon SageMaker Ground Truth classification labeling job. \n",

File: sagemaker-featurestore/feature_store_object_detection_ground_truth.ipynb [hunk 13]
```diff
@@ -0,0 +1,484 @@
...
+    "    \n",
+    "    Return a dataframe containing all the information from your output.manifest file. \n",
```
Comment (jsspric) [inline sagemaker-featurestore/feature_store_object_detection_ground_truth.ipynb]: "    Return a data frame containing all the information from your `output.manifest` file. \n",

File: sagemaker-featurestore/feature_store_object_detection_ground_truth.ipynb [hunk 14]
```diff
@@ -0,0 +1,484 @@
...
+    "                        \"creation_date\": creation_date})\n",
+    "\n",
+    "df = create_dataframe_from_manifest('output.manifest')"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### Preview the parsed manifest file as a dataframe"
```
Comment (jsspric) [inline sagemaker-featurestore/feature_store_object_detection_ground_truth.ipynb]: "#### Preview the parsed manifest file as a data frame"

File: sagemaker-featurestore/feature_store_object_detection_ground_truth.ipynb [hunk 15]
```diff
@@ -0,0 +1,484 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def cast_object_to_string(data_frame):\n",
+    "    \"\"\"\n",
+    "    data_frame: A pandas Dataframe \n",
```
Comment (jsspric) [inline sagemaker-featurestore/feature_store_object_detection_ground_truth.ipynb]: "    data_frame: A `pandas` data frame \n",

File: sagemaker-featurestore/feature_store_object_detection_ground_truth.ipynb [hunk 16]
```diff
@@ -0,0 +1,484 @@
...
+   "outputs": [],
+   "source": [
+    "df"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Create Feature Group and Ingest data into it\n",
```
Comment (jsspric) [inline sagemaker-featurestore/feature_store_object_detection_ground_truth.ipynb]: "### Create a feature group and ingest data into it\n",

File: sagemaker-featurestore/feature_store_object_detection_ground_truth.ipynb [hunk 17]
```diff
@@ -0,0 +1,484 @@
...
+    "\n",
+    "Below we start by appending the `EventTime` feature to your data to timestamp entries, then we load the feature definition, and instantiate the Feature Group object. Then lastly we ingest the data into your feature store."
```
Comment (jsspric) [inline sagemaker-featurestore/feature_store_object_detection_ground_truth.ipynb:245]: "The following example starts by appending the `EventTime` feature to your data to timestamp entries, then loads the feature definition and instantiates the feature group object. Finally, it ingests the data into your feature store."

File: sagemaker-featurestore/feature_store_object_detection_ground_truth.ipynb [hunk 18]
```diff
@@ -0,0 +1,484 @@
...
+   "outputs": [],
+   "source": [
+    "df"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Load Feature Definition's of your data into your feature group."
```
Comment (jsspric) [inline sagemaker-featurestore/feature_store_object_detection_ground_truth.ipynb:316]: "Load feature definitions of your data into your feature group."

File: sagemaker-featurestore/feature_store_object_detection_ground_truth.ipynb [hunk 19]
```diff
@@ -0,0 +1,484 @@
...
+    "feature_group.load_feature_definitions(data_frame=df)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Create your feature group.\n",
+    "\n",
+    "**Important**: You will need to substitute your KMS Key ARN for `kms_key` for server side encryption. "
```
Comment (jsspric) [inline sagemaker-featurestore/feature_store_object_detection_ground_truth.ipynb:334]: "**Important**: You must substitute your KMS key ARN for `kms_key` for server-side encryption. "

File: sagemaker-featurestore/feature_store_object_detection_ground_truth.ipynb [hunk 20]
```diff
@@ -0,0 +1,484 @@
...
+    "    \n",
+    "print('Data available.')"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Clean Up Resources\n",
+    "Remove the Feature Group that was created. "
```
Comment (jsspric) [inline sagemaker-featurestore/feature_store_object_detection_ground_truth.ipynb:449]: "Remove the feature group that was created. "

File: sagemaker-featurestore/feature_store_object_detection_ground_truth.ipynb [hunk 21]
```diff
@@ -0,0 +1,484 @@
...
+    "        time.sleep(60)\n",
+    "    \n",
+    "print('Data available.')"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Clean Up Resources\n",
```
Comment (jsspric) [inline sagemaker-featurestore/feature_store_object_detection_ground_truth.ipynb]: "### Clean up resources\n",

File: sagemaker-featurestore/feature_store_object_detection_ground_truth.ipynb [hunk 22]
```diff
@@ -0,0 +1,484 @@
...
+   "outputs": [],
+   "source": [
+    "feature_group.delete()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Next Steps\n",
```
Comment (jsspric) [inline sagemaker-featurestore/feature_store_object_detection_ground_truth.ipynb]: "### Next steps\n",

File: sagemaker-featurestore/feature_store_object_detection_ground_truth.ipynb [hunk 23]
```diff
@@ -0,0 +1,484 @@
...
+    "\n",
+    "In this notebook we covered how to securely store the output of an image or text classification labelling job from [Amazon Ground Truth](https://docs.aws.amazon.com/sagemaker/latest/dg/sms.html) directly into Feature Store using KMS key.\n",
```
Comment (jsspric) [inline sagemaker-featurestore/feature_store_object_detection_ground_truth.ipynb:467]: "In this notebook, we covered how to securely store the output of an image or text classification labeling job from [Amazon SageMaker Ground Truth](https://docs.aws.amazon.com/sagemaker/latest/dg/sms.html) directly into Feature Store using a KMS key.\n",

File: sagemaker-featurestore/feature_store_object_detection_ground_truth.ipynb [hunk 24]
```diff
@@ -0,0 +1,484 @@
...
+    "\n",
+    "If you are interested in understanding more on how server-side encryption is done with Feature Store, see [Feature Store: Encrypt Data in your Online or Offline Feature Store using KMS key](https://sagemaker-examples.readthedocs.io/en/latest/sagemaker-featurestore/feature_store_kms_key_encryption.html).\n",
```
Comment (jsspric) [inline sagemaker-featurestore/feature_store_object_detection_ground_truth.ipynb:469]: "To learn more about how server-side encryption is done with Feature Store, see [Feature Store: Encrypt Data in your Online or Offline Feature Store using KMS key](https://sagemaker-examples.readthedocs.io/en/latest/sagemaker-featurestore/feature_store_kms_key_encryption.html).\n",

File: sagemaker-featurestore/feature_store_object_detection_ground_truth.ipynb [hunk 25]
```diff
@@ -0,0 +1,484 @@
...
+    "\n",
+    "If you are interested in understanding how to do client-side encryption to encrypt your image data set prior to storing it in your feature store, see [Amazon SageMaker Feature Store: Client-side Encryption using AWS Encryption SDK](https://sagemaker-examples.readthedocs.io/en/latest/sagemaker-featurestore/feature_store_client_side_encryption.html). For more information on the AWS Encryption library, see [AWS Encryption SDK library](https://docs.aws.amazon.com/encryption-sdk/latest/developer-guide/introduction.html).\n",
```
Comment (jsspric) [inline sagemaker-featurestore/feature_store_object_detection_ground_truth.ipynb:471]: "To learn more about how to do client-side encryption to encrypt your image dataset prior to storing it in your feature store, see [Amazon SageMaker Feature Store: Client-side Encryption using AWS Encryption SDK](https://sagemaker-examples.readthedocs.io/en/latest/sagemaker-featurestore/feature_store_client_side_encryption.html). For more information on the AWS Encryption library, see [AWS Encryption SDK library](https://docs.aws.amazon.com/encryption-sdk/latest/developer-guide/introduction.html).\n",

**Review by can-sun — APPROVED** [review]

**Pull request merged**